File: electron/src/workshop.ts
// SPDX-License-Identifier: PolyForm-Noncommercial-1.0.0
// STEAM WORKSHOP — where a player's mods come from, and where an author's goes.
//
// The seam's third file, in the same shape as cloud save's and the
// achievements': a bridge above it moves JSON, and this is the only module that
// knows Steam exists. That is what keeps the web side from ever learning which
// platform answered — and why the day a second storefront grows a mod portal,
// it is one new file here rather than a change to the protocol.
//
// Two directions, and they are asymmetric on purpose:
//
//   SUBSCRIBE  Steam does all of it. The client downloads a subscribed item
//              into its own folder and we ask where that folder is. There is
//              no install step of ours to get wrong, and no unpacking — which
//              also means no archive parser pointed at a stranger's file.
//   PUBLISH    We hand Steam a FOLDER and it uploads the contents. So a mod is
//              published exactly as authored: the YAML a human wrote, not the
//              compiled bundle. The subscriber's game compiles it locally, and
//              a mod on the Workshop stays readable, forkable and diffable the
//              way the game's own content is.
//
// The whole module degrades to "no Workshop" without Steam: `steamClient()`
// answers null on a developer machine, in CI, and on a build launched outside
// the client, and every function here returns an empty list or a refusal
// rather than throwing. A game with no mods is the game.

import { existsSync } from "node:fs";
import path from "node:path";

import { output } from "./output";
import {
  ITEM_STATE_INSTALLED,
  ITEM_STATE_NEEDS_UPDATE,
  steamClient,
} from "./steam";

/** A mod as Steam knows it: an id and a folder on disk. Nothing is read or
 * validated here — that is the compiler's job, one layer up. */
export type WorkshopItem = {
  /** The published file id, as a string: it is a uint64 and JSON has no such
   * number, so it travels as text the whole way to the page and back. */
  itemId: string;
  folder: string;
  needsUpdate: boolean;
};

/** Every installed, subscribed Workshop item. */
export function subscribedItems(): WorkshopItem[] {
  const client = steamClient();
  if (!client) return [];

  const items: WorkshopItem[] = [];
  let ids: bigint[];
  try {
    ids = client.workshop.getSubscribedItems();
  } catch (err) {
    output.warn(`workshop: could not list subscriptions — ${describe(err)}`);
    return [];
  }

  for (const id of ids) {
    let state = 0;
    try {
      state = client.workshop.state(id);
    } catch {
      continue;
    }
    // A subscription the client has not finished downloading yet is not a
    // failure and not a mod — it is a download in progress. Kick it along and
    // leave it out of this pass; the next launch (or the next refresh) sees it.
    if ((state & ITEM_STATE_INSTALLED) === 0) {
      try {
        client.workshop.download(id, false);
      } catch {
        /* the client will get to it */
      }
      continue;
    }
    const info = safeInstallInfo(client, id);
    if (!info || !existsSync(info.folder)) continue;
    items.push({
      itemId: id.toString(),
      folder: info.folder,
      needsUpdate: (state & ITEM_STATE_NEEDS_UPDATE) !== 0,
    });
  }
  return items;
}

/** What a publish attempt answers with. `agreement` is its own outcome because
 * it is the one failure the player must go and DO something about: Steam
 * refuses to show an item until its author has accepted the Workshop terms in
 * a browser, and the item exists in the meantime, invisible. */
export type PublishResult =
  | { ok: true; itemId: string; needsToAcceptAgreement: boolean }
  | { ok: false; reason: "no-steam" | "not-a-mod" | "error"; detail?: string };

/**
 * Publish a mod folder, creating the Workshop item on the first call.
 *
 * `itemId` is null the first time and the returned id every time after — the
 * game remembers it in the mod's own folder (`.workshop-id`), so "publish an
 * update" is the same command as "publish", and a mod cannot accidentally
 * become two Workshop entries.
 */
export async function publishMod(
  folder: string,
  meta: {
    itemId: string | null;
    title: string;
    description: string;
    changeNote: string;
    previewPath?: string;
    tags?: string[];
  },
): Promise<PublishResult> {
  const client = steamClient();
  if (!client) return { ok: false, reason: "no-steam" };
  if (!existsSync(path.join(folder, "mod.yaml"))) {
    return { ok: false, reason: "not-a-mod" };
  }

  try {
    const itemId =
      meta.itemId !== null
        ? BigInt(meta.itemId)
        : (await client.workshop.createItem()).itemId;

    const result = await client.workshop.updateItem(itemId, {
      title: meta.title,
      description: meta.description,
      changeNote: meta.changeNote,
      // The AUTHORED folder, not a compiled bundle — see the header.
      contentPath: folder,
      ...(meta.previewPath && existsSync(meta.previewPath)
        ? { previewPath: meta.previewPath }
        : {}),
      ...(meta.tags?.length ? { tags: meta.tags } : {}),
    });
    output.info(`workshop: published ${meta.title} as ${result.itemId}`);
    return {
      ok: true,
      itemId: result.itemId.toString(),
      needsToAcceptAgreement: result.needsToAcceptAgreement,
    };
  } catch (err) {
    output.warn(`workshop: publish failed — ${describe(err)}`);
    return { ok: false, reason: "error", detail: describe(err) };
  }
}

function safeInstallInfo(
  client: NonNullable<ReturnType<typeof steamClient>>,
  id: bigint,
): { folder: string } | null {
  try {
    return client.workshop.installInfo(id);
  } catch {
    return null;
  }
}

const describe = (err: unknown): string =>
  err instanceof Error ? err.message : String(err);

File: tauri/shell/src/workshop.rs
// SPDX-License-Identifier: PolyForm-Noncommercial-1.0.0
//! STEAM WORKSHOP — where a player's mods come from, and where an author's
//! goes. The peer of `electron/src/workshop.ts`, and the seam's third file in
//! the same shape as cloud save's and the achievements': a bridge above it
//! moves JSON, this declares what a mod portal has to be, and only
//! `src-tauri/src/workshop.rs` knows Steam exists.
//!
//! That is what keeps the web side from ever learning which platform answered —
//! and why the day a second storefront grows a mod portal, it is one new file
//! there rather than a change to the protocol.
//!
//! Two directions, and they are asymmetric on purpose:
//!
//!   SUBSCRIBE  Steam does all of it. The client downloads a subscribed item
//!              into its own folder and we ask where that folder is. There is
//!              no install step of ours to get wrong, and no unpacking — which
//!              also means no archive parser pointed at a stranger's file.
//!   PUBLISH    We hand Steam a FOLDER and it uploads the contents. So a mod is
//!              published exactly as authored: the YAML a human wrote, not the
//!              compiled bundle. The subscriber's game compiles it locally, and
//!              a mod on the Workshop stays readable, forkable and diffable the
//!              way the game's own content is.
//!
//! The whole seam degrades to "no Workshop" without Steam: the provider is
//! `None` on a developer machine, in CI, and on a build launched outside the
//! client, and every path here answers an empty list or a refusal rather than
//! an error. A game with no mods is the game.

/// A mod as Steam knows it: an id and a folder on disk.
///
/// Nothing is read or validated here — that is the compiler's job, one layer
/// up.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct WorkshopItem {
    /// The published file id, as a string: it is a uint64 and JSON has no such
    /// number, so it travels as text the whole way to the page and back.
    pub item_id: String,
    /// Where the client put it.
    pub folder: String,
    /// Steam has a newer version than the one on disk.
    pub needs_update: bool,
}

/// What a publish needs to know about the mod, read off its compiled bundle.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct PublishRequest {
    /// The item to update, or `None` to create one.
    pub item_id: Option<String>,
    /// The mod's display name.
    pub title: String,
    /// Its description.
    pub description: String,
    /// What changed, as the player typed it.
    pub change_note: String,
    /// The AUTHORED folder, not a compiled bundle — see the module header.
    pub folder: String,
    /// A thumbnail, when the mod ships one.
    pub preview: Option<String>,
    /// The Workshop tags, from the mod's own kind.
    pub tags: Vec<String>,
}

/// Which Workshop tag a mod's `kind` earns.
///
/// Two, because two is what the Workshop's own filter is worth: a player
/// browsing wants to know whether this replaces the game or adds to it, and any
/// finer taxonomy is a set of tags nobody selects.
pub fn tag_for_kind(kind: &str) -> &'static str {
    if kind == "conversion" {
        "Total Conversion"
    } else {
        "Addon"
    }
}

/// Steam's own item-state bits, as much of them as this shell reads.
pub mod state {
    /// The client has finished downloading it.
    pub const INSTALLED: u32 = 4;
    /// …and a newer version exists.
    pub const NEEDS_UPDATE: u32 = 8;
}

/// What to do with one subscribed item, from its state bits.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum ItemDisposition {
    /// It is on disk and ready to compile.
    Ready {
        /// Whether Steam has a newer version.
        needs_update: bool,
    },
    /// The client has not finished downloading it. Kick the download along and
    /// leave it out of this pass — the next launch (or the next refresh) sees
    /// it. A subscription mid-download is not a failure and not a mod.
    Downloading,
}

/// Read one item's state bits.
pub fn disposition(bits: u32) -> ItemDisposition {
    if bits & state::INSTALLED == 0 {
        return ItemDisposition::Downloading;
    }
    ItemDisposition::Ready {
        needs_update: bits & state::NEEDS_UPDATE != 0,
    }
}

/// What a publish attempt answered with, before the bridge dresses it.
#[derive(Debug, Clone, PartialEq, Eq)]
pub enum PublishAnswer {
    /// It went up.
    Ok {
        /// The item it went up as.
        item_id: String,
        /// Whether the author still has to accept the Workshop terms.
        needs_to_accept_agreement: bool,
    },
    /// It did not.
    Failed {
        /// What Steam said.
        detail: String,
    },
}

/// Somewhere a mod can be subscribed to and published — the platform seam.
///
/// Synchronous for the reason the cloud seam is: Steam's calls return on the
/// calling thread and this shell has no async runtime to promise into. The one
/// exception on the real client is the UPLOAD, which the implementation blocks
/// on inside `src-tauri/src/workshop.rs` — a publish is already a long press on
/// a button the page has a ten-minute timeout for.
pub trait WorkshopProvider: Send + Sync {
    /// Every installed, subscribed item.
    fn subscribed(&self) -> Vec<WorkshopItem>;
    /// Publish (or update) one folder.
    fn publish(&self, request: &PublishRequest) -> PublishAnswer;
}

/// The line the shell logs when a mod could not be read at all.
///
/// The compiler THROWING rather than reporting is a bug in US, not in the mod —
/// but it must still not take the list down with it, so the row appears with
/// this as its error.
pub fn compiler_failed(key: &str, detail: &str) -> String {
    format!("mods: {key} could not be read — {detail}")
}
